app/src/main/java/uk/co/neuralcubes/neuralates/ControlFragment.java

```diff
@@ -118,72 +126,119 @@ public void onClick(View v) {
+            mMuseHandler.get().connect(mBus);
 
-            public void onNothingSelected(AdapterView<?> parent) {
-                // Another interface callback
-                setEnabledStateForViews(mSpheroActions, false);
+            setEnabledStateForViews(mMuseActions, true);
+
+            if (mSphero.isPresent()) {
+                mController = Optional.of(new RobotController(mSphero.get(), mBus));
             }
-        });
+        }
+    }
 
-        mSelectMuse.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
-            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
+    public void onNothingSelected(AdapterView<?> adapterView) {
+        if (adapterView == mSelectSphero) {
+            setEnabledStateForViews(mSpheroActions, false);
+            if (mSphero.isPresent()) {
+                mSphero.get().removeResponseListener(this);
+                updateSpheroBatteryLevel(null);
+            }
+        } else if (adapterView == mSelectMuse) {
+            setEnabledStateForViews(mMuseActions, false);
+            if (mController.isPresent()) {
+                mController.get().setOverrideFocus(false);
+            }
+        }
+    }
 
-                //0 is the default "Choose muse" element in the spinner
-                if (i > 0) {
-                    //fix the offset
-                    mMuseHandler = Optional.of(PairedMuse.getPairedMuses().get(i - 1));
-                    mMuseHandler.get().connect(mBus);
+    // END - AdapterView.OnItemSelectedListener
 
-                    setEnabledStateForViews(mMuseActions, true);
+    // BEGINNING - Sphero's ResponseListener
 
-                    if (mSphero.isPresent()) {
-                        mController = Optional.of(new RobotController(mSphero.get(), mBus));
-                    }
-                }
-            }
+    @Override
+    public void handleResponse(DeviceResponse deviceResponse, Robot robot) {
+        Log.d("SPHERO", "handleResponse" + deviceResponse.toString());
+        if (deviceResponse instanceof GetPowerStateResponse) {
+            updateSpheroBatteryLevel(((GetPowerStateResponse)deviceResponse).getPowerState());
+        }
+    }
 
-            public void onNothingSelected(AdapterView<?> parent) {
-                setEnabledStateForViews(mMuseActions, false);
-                if (mController.isPresent()) {
-                    mController.get().setOverrideFocus(false);
-                }
-            }
-        });
+    @Override
+    public void handleStringResponse(String s, Robot robot) {
+        Log.d("SPHERO", "handleStringResponse" + s.toString());
+    }
+
+    @Override
+    public void handleAsyncMessage(AsyncMessage asyncMessage, Robot robot) {
+        Log.d("SPHERO", "handleAsyncMessage" + asyncMessage.toString());
+    }
+
+    // END - Sphero's ResponseListener
+
+    private void updateSpheroBatteryLevel(GetPowerStateResponse.PowerState powerState) {
```

Comment:
Shouldn't this or the call to this method be run on the UI thread? I presume libsphero will call this concurrently, not triggered from the ui.